packages/apollo-server-core/src/errors.ts

```diff
@@ -199,6 +199,28 @@ export class ForbiddenError extends ApolloError {
   }
 }
 
+export class BadUserInputError extends ApolloError {
+  constructor(message: string, properties?: Record<string, any>) {
+    super(message, 'BAD_USER_INPUT', properties);
+
+    // Set the prototype explicitly.
+    // https://stackoverflow.com/a/41102306
+    Object.setPrototypeOf(this, BadUserInputError.prototype);
+    Object.defineProperty(this, 'name', { value: 'BadUserInputError' });
+  }
+}
+
+export class TransientError extends ApolloError {
```

Comment:
Let's leave the `TransientError` out for now, since we don't have a compelling use case on the client-side. It seems like the best place to handle theses sorts of errors is either in the layer above the GraphQL execution, such as middleware, or in the Data sources, which could retry on their own. I totally agree that we want this sort of retriability. I'm just not convinced that an error in the resolvers is the right place.

Reply:
I agree with you. Let's keep it as a suggested pattern so that our use cases for ApolloError stays solid!

Reply:
Working updating the error handling section of the docs as well. What do you feel about moving some of the code examples from my blog over as well once they have been reviewed?

Reply:
Totally! We can move them over now too if you prefer